src/main/java/jenkins/scm/api/SCMSource.java

```diff
@@ -197,10 +199,17 @@ public final synchronized boolean hasId() {
      */
     @NonNull
     public final synchronized String getId() {
-        if (id == null) {
-            id = UUID.randomUUID().toString();
+        if (this.id == null) {
+            this.id = UUID.randomUUID().toString();
+            if (owner != null) {
```

Comment:
Ahhh... I wonder if we shouldn’t just assign the id during set owner instead

Reply:
*shrug*

Reply:
That was a request to try not an idle comment ;-)

Reply:
Well, my *shrug* was at least in part that I don't know if that's the right approach - it's entirely possible for someone to call `setId(String)` after `setOwner(..)` is called but before `getId()` is called.

Reply:
Once setOwner is called with non-null then the ID must be assigned if not already imho.

Now the question then becomes will the ID be assigned if saved *before* setOwner... and that is the rub...

May need to add a protected writeReplace that does the assignment... but then the test needs a pair of subclasses... one overriding and one not... or perhaps an XStream converter

Reply:
I legitimately believe that the current implementation is the best (or at least least worst) yet discussed. It doesn't change any existing expectations (as autogenerating the ID on `setOwner(...)` if `id` is `null` would), it works fine with calling `setId(...)` or `getId()` before `setOwner(...)` is called, etc. Least complicated option is best option.

Reply:
Why not leave this here, but also add a `transient` flag indicating that the ID had been generated; and then in `setOwner`, if that flag is set, save the owner then also? I think that covers all the bases without getting too complicated. OTOH it seems likely that a `setOwner` call would have already been accompanied by a save.